lib/auth/gssapi.js

```diff
@@ -1,114 +1,45 @@
-};
-
-AuthSession.prototype.equal = function(session) {
-  return (
-    session.db === this.db &&
-    session.username === this.username &&
-    session.password === this.password
-  );
-};
+const AuthProvider = require('./authProvider').AuthProvider;
 
 /**
  * Creates a new GSSAPI authentication mechanism
  * @class
- * @return {GSSAPI} A cursor instance
- */
-var GSSAPI = function(bson) {
-  this.bson = bson;
-  this.authStore = [];
-};
-
-/**
- * Authenticate
- * @method
- * @param {{Server}|{ReplSet}|{Mongos}} server Topology the authentication method is being called on
- * @param {[]Connections} connections Connections to authenticate using this authenticator
- * @param {string} db Name of the database
- * @param {string} username Username
- * @param {string} password Password
- * @param {authResultCallback} callback The callback to return the result from the authentication
- * @return {object}
+ * @extends AuthProvider
  */
-GSSAPI.prototype.auth = function(server, connections, db, username, password, options, callback) {
-  var self = this;
-  // We don't have the Kerberos library
-  if (Kerberos == null) return callback(new Error('Kerberos library is not installed'));
-  var gssapiServiceName = options['gssapiServiceName'] || 'mongodb';
-  // Total connections
-  var count = connections.length;
-  if (count === 0) return callback(null, null);
-
-  // Valid connections
-  var numberOfValidConnections = 0;
-  var errorObject = null;
-
-  // For each connection we need to authenticate
-  while (connections.length > 0) {
-    // Execute MongoCR
-    var execute = function(connection) {
-      // Start Auth process for a connection
-      GSSAPIInitialize(
-        self,
-        db,
-        username,
-        password,
-        db,
-        gssapiServiceName,
-        server,
-        connection,
-        options,
-        function(err, r) {
-          // Adjust count
-          count = count - 1;
-
-          // If we have an error
-          if (err) {
-            errorObject = err;
-          } else if (r.result['$err']) {
-            errorObject = r.result;
-          } else if (r.result['errmsg']) {
-            errorObject = r.result;
-          } else {
-            numberOfValidConnections = numberOfValidConnections + 1;
-          }
-
-          // We have authenticated all connections
-          if (count === 0 && numberOfValidConnections > 0) {
-            // Store the auth details
-            addAuthSession(self.authStore, new AuthSession(db, username, password, options));
-            // Return correct authentication
-            callback(null, true);
-          } else if (count === 0) {
-            if (errorObject == null)
-              errorObject = new MongoError(f('failed to authenticate using mongocr'));
-            callback(errorObject, false);
-          }
-        }
-      );
-    };
+class GSSAPI extends AuthProvider {
+  _authenticateSingleConnection(server, connection, credentials, callback) {
```

Comment:
Can you document this as an `@override`?

Reply:
On a reread, this is actually a method that `AuthProvider` _assumes_ exists, but is not actually defined with a default implementation. I recommend adding a default imply that throws an error